security/signing.py

```diff
@@ -0,0 +1,140 @@
+
+
+class MessageVerifier:
+    def get_public_keys(self):
+        r = get_redis()
+        public_keys = {}
+        s = requests.Session()
+        all_public_keys = PublicKey.objects.filter(is_active=1)
+        for current_public_key in all_public_keys:
+            redis_jwks_key = str(current_public_key.id) + "-jwks"
+            current_kid = current_public_key.key_id
+            if r.exists(redis_jwks_key):
+                k = r.get(redis_jwks_key)
+                key = json.loads(k)
+            else:
+                response = s.get(current_public_key.url)
+                jwks_data = response.json()
+                if "keys" in jwks_data:
+                    jwk = next(
+                        (
+                            item
+                            for item in jwks_data["keys"]
+                            if item["kid"] == current_kid
+                        ),
+                        None,
+                    )
+                else:
+                    if "kid" in jwks_data.keys():
+                        jwk = jwks_data if current_kid == jwks_data["kid"] else None
+                    else:
+                        jwk = None
+                key = jwk if jwk else {"000"}
+
+                r.set(redis_jwks_key, json.dumps(key))
+                r.expire(redis_jwks_key, 60000)
+            public_keys[current_kid] = key
+        return public_keys
+
+    def verify_message(self, request) -> bool:
+        stored_public_keys = self.get_public_keys()
+        if bool(not stored_public_keys):
+            return False
+
+        r = requests.Request(
+            request.method,
+            request.build_absolute_uri(),
+            json=request.data,
+            headers=request.headers,
+        )
+
+        for _, json_web_token in stored_public_keys.items():
+            verifier = HTTPMessageVerifier(
+                signature_algorithm=algorithms.RSA_PSS_SHA512,
+                key_resolver=BlenderHTTPSignatureKeyResolver(jwk=json_web_token),
+            )
+            verifier.verify(r)
+        return True
+
+
+class ResponseSigningOperations:
+    def __init__(self):
+        self.signing_url = env.get("FLIGHT_PASSPORT_SIGNING_URL", None)
+        self.signing_client_id = env.get("FLIGHT_PASSPORT_SIGNING_CLIENT_ID")
+        self.signing_client_secret = env.get("FLIGHT_PASSPORT_SIGNING_CLIENT_SECRET")
+
+    def generate_content_digest(self, payload):
+        payload_str = json.dumps(payload)
+        return str(
+            http_sfv.Dictionary(
+                {"sha-256": hashlib.sha256(payload_str.encode("utf-8")).digest()}
+            )
+        )
+
+    def sign_json_via_django(self, data_to_sign):
+        signer = Signer()
+        signed_obj = signer.sign_object(data_to_sign)
+        return signed_obj
+
+    def sign_json_via_jose(self, payload):
+        """
+        For a payload sign using the OIDC private key and return signed JWS
+        """
+
+        algorithms = "RS256"
```

Comment:
Lets keep hard coded attributes as class member, to be assigned with current default from constructor and later to be overridden from env

Reply:
Done [961924d](https://github.com/tiiuae/flight-blender/pull/28/commits/961924da618fa397025c49e5df26307ff91b525e)